src/pagination.rs

```diff
@@ -0,0 +1,273 @@
+        } else {
+          Ordering::Greater
+        }
+      } else {
+        Ordering::Less
+      };
+      if cmp != Ordering::Equal {
+        return cmp;
+      }
+    }
+    cmp
+  });
+}
+
+fn create_all_paginators(
+  mut all_posts: &mut Vec<liquid::Value>,
+  doc: &Document,
+  pagination_cfg: &PaginationConfig,
+) -> Vec<liquid::Object> {
+  let mut paginators = Vec::new();
+  let total_posts = all_posts.len() as i32;
+  let total_pages = (total_posts as f32 / pagination_cfg.per_page as f32).ceil() as i32;
+  for i in 0..total_pages {
+    paginators.push(create_paginator(
+      i,
+      total_pages,
+      total_posts,
+      &pagination_cfg,
+      &doc,
+      &mut all_posts,
+    ));
+  }
+  paginators
+}
+
+fn create_paginator(
+  i: i32,
+  total_pages: i32,
+  total_posts: i32,
+  config: &PaginationConfig,
+  doc: &Document,
+  mut all_posts: &mut Vec<liquid::Value>,
+) -> liquid::Object {
+  let page = i + 1;
+  let mut paginator = liquid::Object::new();
+  let file_name = doc
+    .file_path
+    .file_name()
+    .map_or("index.html", |s| s.to_str().unwrap_or("index.html"));
+
+  init_paginator_constants(
+    &mut paginator,
+    total_posts,
+    total_pages,
+    &config,
+    &file_name,
+    &doc,
+  );
+
+  fill_current_page_info(
+    &mut paginator,
+    page,
+    &mut all_posts,
+    &config,
+    &file_name,
+    &doc,
+  );
+
+  fill_previous_next_info(&mut paginator, page, total_pages, &file_name, &doc, &config);
+
+  paginator
+}
+
+fn pagination_attributes(page_num: i32, include: Include) -> liquid::Object {
+  let mut attributes = liquid::Object::new();
+  attributes.insert("num".to_owned(), liquid::Value::scalar(page_num));
+  attributes.insert("include".to_owned(), liquid::Value::scalar(include));
+  attributes
+}
+
+fn interpret_permalink(
+  config: &PaginationConfig,
+  doc: &Document,
+  page_num: i32,
+  file_name: &str,
+) -> String {
+  let cfg = config.clone();
+  let mut attributes = document::permalink_attributes(&doc.front, &doc.file_path);
+  let pagination_attr = pagination_attributes(page_num, cfg.include);
+  pagination_attr.into_iter().for_each(|(k, v)| {
+    attributes.insert(k, v);
+  });
+  let permalink = permalink::explode_permalink(&config.permalink, &attributes);
+  let p = permalink
+    .and_then(|mut p| {
+      p.push_str(file_name);
+      Ok(p)
+    })
+    .unwrap_or(file_name.to_owned());
+  p
+}
+
+fn init_paginator_constants(
+  paginator: &mut liquid::Object,
+  total_posts: i32,
+  total_pages: i32,
+  config: &PaginationConfig,
+  file_name: &str,
+  doc: &Document,
+) {
+  paginator.insert(
+    "per_page".to_owned(),
+    liquid::Value::scalar(config.per_page),
+  );
+  paginator.insert("total_posts".to_owned(), liquid::Value::scalar(total_posts));
+  paginator.insert("total_pages".to_owned(), liquid::Value::scalar(total_pages));
+  paginator.insert(
+    "first_page_path".to_owned(),
+    liquid::Value::scalar(format!("/{}", doc.file_path.to_string_lossy())),
```

Comment:
Why use `file_path` instead of `url_path`?

Reply:
I've overlook `url_path`, it may be more adapted, I'll have a look